File: src/lessons/welcome/loopdowhile/PoucetEntity.java
package lessons.welcome.loopdowhile;

import java.awt.Color;

import plm.core.model.Game;

public class PoucetEntity extends plm.universe.bugglequest.SimpleBuggle {
	@Override
	public void forward(int i)  { 
		throw new RuntimeException(Game.i18n.tr("I'm sorry Dave, I'm affraid I can't let you use forward with an argument."));
	}
	@Override
	public void backward(int i) {
		throw new RuntimeException(Game.i18n.tr("I'm sorry Dave, I'm affraid I can't let you use backward with an argument."));
	}
	
	public boolean crossing() {
		return getX()%5== 1 && getY()%5==1;
	}
	public boolean exitReached() {
		return getGroundColor().equals(Color.orange);
	}	
	/* BINDINGS TRANSLATION */
	boolean sortieTrouvee() { return exitReached(); }
	boolean croisement() { return crossing(); }

	

	@Override
	/* BEGIN TEMPLATE */
	public void run() { 
		/* BEGIN SOLUTION */
		while (!exitReached()) {
			int seen = 0;
			
			do {
				forward();
				if (isOverBaggle())
					seen++;
			} while (! crossing());
			
			if (seen>2)
				left();
			else
				right();
		}
		forward();
		/* END SOLUTION */
	}
	/* END TEMPLATE */
}

File: src/lessons/welcome/loopdowhile/PoucetEntity.py
def forward(i=-1):
    if i==-1:
      entity.forward()
    else:
      errorMsg("Sorry Dave, I cannot let you use forward with an argument in this exercise. Use a loop instead.")
def backward(i=-1):
    if i==-1:
      entity.backward()
    else:
      errorMsg("Sorry Dave, I cannot let you use backward with an argument in this exercise. Use a loop instead.")

def crossing():
    return (entity.getX() % 5 == 1) and ( entity.getY()%5==1 )
    
def exitReached():
    return getGroundColor().equals(Color.orange)

# BINDINGS TRANSLATION 
def croisement():
    return crossing()
def sortieTrouvee():
    return exitReached()


# BEGIN SOLUTION
while not exitReached() :
    seen = 0
    within = False
    
    while not within or not crossing():
        within = True
        forward()
        if isOverBaggle():
            seen += 1
    
    if seen > 2:
        left()
    else:
        right()
forward()
# END SOLUTION
